src/main/java/org/graylog/collector/file/FileObserver.java

```diff
@@ -103,20 +91,19 @@ public FileObserver(WatchService watchService) {
         watcher = watchService;
     }
 
-    public void observePath(Listener listener, Path path, FileNamingStrategy namingStrategy) throws IOException {
+    public void observePathSet(PathSet pathSet, Listener listener) throws IOException {
         Preconditions.checkNotNull(listener);
-        Preconditions.checkNotNull(path);
-        Preconditions.checkNotNull(namingStrategy);
+        Preconditions.checkNotNull(pathSet);
 
-        Path directory = path;
-        if (!Files.isDirectory(directory, LinkOption.NOFOLLOW_LINKS)) {
-            directory = path.getParent();
-        }
-        final WatchKey key = directory.register(watcher, new WatchEvent.Kind[]{ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY},
+        final Path rootPath = pathSet.getRootPath();
+
+        log.debug("Watching directory {} for changes matching: {}", rootPath, pathSet.getPattern());
+
+        final WatchKey key = rootPath.register(watcher, new WatchEvent.Kind[]{ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY},
                 SensitivityWatchEventModifier.HIGH);
-        log.debug("Watching directory {} for file {}", directory, path);
-        keys.putIfAbsent(key, new WatchPath(directory));
-        keys.get(key).addEntry(new WatchEntry(path, namingStrategy, listener));
+
+        keys.putIfAbsent(key, Sets.<WatchPath>newConcurrentHashSet());
+        keys.get(key).add(new WatchPath(pathSet, listener));
```

Comment:
Should initializing the key with an empty `ConcurrentHashSet` and adding the actual `WatchPath` be atomic?

If there are multiple threads accessing this object/method, we should guard this initialization and adding the `WatchPath` with a lock (or `synchronized(keys)`) because the thread might be suspended after `keys.get(key)`.